### PR TITLE
Merge HeliFly into Fly and make TakeOff self-contained

### PR DESCRIPTION
**Pull request by reaperrr**
Split from #16483.

Please let's leave any further non-trivial refactors to #16483, which will effectively become the follow-up to this.
The aim of this PR is to get the activity merge into bleed without regressions, so we have a stable base for further improvements (and less time pressure).

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/Air/Fly.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Fly.cs]: This still has the issue where it will do the wrong thing (miss the desiredAltitude and possibly move at angles greater than MaximumPitch) if `move.Z != 0`

File: OpenRA.Mods.Common/Activities/Air/Fly.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Fly.cs]: Did you mean `var desiredFacing = aircraft.Info.CanHover && delta.HorizontalLengthSquared != 0 ? delta.Yaw.Facing : aircraft.Facing;`?

The current line will always return `delta.Yaw.Facing` for non-hovering aircraft, even if `delta` is zero, meaning the yaw is bogus.

File: OpenRA.Mods.Common/Activities/Air/Fly.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Fly.cs]: Can `aircraft.Info.CruiseAltitude` and `targetAltitude` be very different here? If so, this will make it jump instantly between positions. Why not call `VerticalTakeOffOrLandTick`?

File: OpenRA.Mods.Common/Activities/Air/TakeOff.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/TakeOff.cs]: Shouldn't this be done from `OnFirstRun`? Otherwise we will double-tick this if we don't make it above `LandAltitude` by the second tick.

File: OpenRA.Mods.Common/Activities/Air/Fly.cs
Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Air/Fly.cs]: I meant it the way I wrote it, because I took that from the old
https://github.com/OpenRA/OpenRA/pull/16584/commits/78c930c94b0e4709dbde2a5b16f3644d6ab1fb57#diff-558171f0691cfeefaa697d6b9dd0734bL126

I assume `var desiredFacing = delta.HorizontalLengthSquared != 0 ? delta.Yaw.Facing : aircraft.Facing;` would be the correct fix then.

**Comment by reaperrr** [comment]
Updated.

**Review by pchote — DISMISSED** [review]

**Comment by pchote** [comment]
The code changes look broadly fine, but I found two regressions while testing:

* If you move a helicopter over a terrain height change in TS (from bottom to top of a cliff works best) and then stop it as soon as it reaches the top it will stay hovering at its current altitude, which is much lower than the standard hovering altitude.  This is less bad than the current behaviour on bleed (stopping until it reaches the right height), so I would be okay with merging without fixing this if the fix is going to be complicated
* Issuing a fly order while a helicopter is part way through landing will make it take off diagonally, much like the issues I noted in #16483.

File: mods/d2k/rules/aircraft.yaml
Comment (pchote) [inline mods/d2k/rules/aircraft.yaml]: IMO it would be best to leave this to #16574 or #16509.

Comment (reaperrr) [reply mods/d2k/rules/aircraft.yaml]: Ok, will remove it.

**Comment by reaperrr** [comment]
Updated.


> If you move a helicopter over a terrain height change in TS (from bottom to top of a cliff works best) and then stop it as soon as it reaches the top it will stay hovering at its current altitude, which is much lower than the standard hovering altitude. This is less bad than the current behaviour on bleed (stopping until it reaches the right height), so I would be okay with merging without fixing this if the fix is going to be complicated

Fixed via `OnBecomingIdle` in `Aircraft`, didn't notice any unwanted side-effects, either.

> Issuing a fly order while a helicopter is part way through landing will make it take off diagonally, much like the issues I noted in #16483.

Fixed by making `Land` uninterruptible.

Additionally,
- made `TakeOff` uninterruptible as well (except on `ForceLanding`)
- clarified descriptions of `MaximumPitch` and `AltitudeVelocity`
- fixed `MaximumPitch` values in TS to make all aircraft (not only Orca bomber and Banshee) adjust to cliffs naturally and not in slow motion.

**Comment by tovl** [comment]
> Fixed by making Land uninterruptible.

We already discussed this in #16509 and found that the loss in responsiveness was undesirable. Furthermore, the implementation here is incompatible with making `Land` self-contained because it would mean that the entire approach would be uninterruptible. A better solution would be to queue `TakeOff` as a childactivity whenever the altitude is lower than cruise-altitude on the first run.

**Comment by reaperrr** [comment]
Updated.

Reverted uninterruptability and instead queue a `TakeOff` child activity in `Land` if `NextActivity != null`. The case of `NextActivity == null` is already handled fully by `OnBecomingIdle` now, thanks to https://github.com/OpenRA/OpenRA/pull/16584/commits/e980a4663ec6c0f3d545e788000efa62354d8c09.

Additionally, added a speculative fix for #16579 (last commit).

**Comment by tovl** [comment]
Actually, I meant to queue it as a childactivity of `Fly` instead of `Land`, but I guess this works just as well.

Speculative fix looks good at first glance.

File: OpenRA.Mods.Common/Activities/Air/Land.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Land.cs]: I think we can skip this check and always do the height check / TakeOff?
Better to always do it here than to rely on two different cases (`Aircraft.OnBecomingIdle`) for idle/not idle handling.

We'll still need the code in Aircraft to handle edge cases like the terrain height changes during flight, but that is fine.

**Comment by pchote** [comment]
> EDIT: On second thought, wouldn't this also trigger when the next activity is another Land activity? If so spam clicking would mean the heli would rise to CruiseAltitude before going back to the ground anyway. Doing it in Fly might therefore be cleaner after all. (This could then also replace the TakeOff queued around line 122 as well)

I like the idea of handling it in `Land` (with my suggested tweak) because it follows the rule that an activity should always leave the actor in a valid state when it completes/cancels.  The cliff-height edge case can then be handled the same way in `Fly`.

Spam clicking is a more general problem that applies to many other activities (#16569) and that is going to be best handled with a more general fix on the `ResolveOrder` side - probably introducing a new standard code pattern to check if the current activity matches the requested order, then skipping the re-queue if they match.

**Comment by tovl** [comment]
I like the idea of catching spam clicks on the `ResolveOrder` side instead of the activity side. However, this would still leave edge-cases where `Land` is a child of another activity. E.g. for chinooks, when a force-land order is directly followed by an unload order.

**Comment by pchote** [comment]
That could perhaps be solved in two steps:
1. Add a new `Land.Cancel` overload that includes a `skipAltitudeFix` flag
2. Add a new `IActivityAircraftAltitudeAware` interface (maybe with a better name), and have the parent activities cancel call `Land.Cancel` with `skipAltitudeFix: NextActivity is IActivityAircraftAltitudeAware`.

My concern is that if we don't handle this at all in `Land` that we will be opening up potentially endless edge cases when other activities that are not explicitly aircraft related are queued after a cancelled Land.

File: OpenRA.Mods.Common/Activities/Air/Land.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: Note that in that case this should be aware of `Aircraft.LandWhenIdle` as well.

Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: > Note that in that case this should be aware of Aircraft.LandWhenIdle as well.

Generally speaking I agree, but I'd like to leave migrating the current `OnBecomingIdle` cases to a dedicated follow-up, as my first attempt to cover this here promptly resulted in some nasty "activity went into a weird uncancellable loop".

This PR already has scope-creeped beyond what I originally wanted it to be, I'd like to leave further improvements to dedicated, purpose-focused follow-ups.

**Comment by reaperrr** [comment]
Updated.

Only https://github.com/OpenRA/OpenRA/pull/16584/commits/0524c3e7c156a27ffe420bf7617f39e41c9704fe changed.

File: OpenRA.Mods.Common/Activities/Air/Land.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Land.cs]: I think this will be best handled by having those activities opt out of the altitude fixup here and handle the taking off themselves (which they are almost certainly already doing).

File: OpenRA.Mods.Common/Activities/Air/Land.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Land.cs]: This is fairly straightforward:
```csharp
if (IsCanceling || target.Type == TargetType.Invalid)
{
	// We must return the actor to a sensible height before continuing
	// If the aircraft lands when idle and is idle, continue landing
	// Otherwise climb back to CruiseAltitude
	// TODO: Allow specific NextActivities to opt out of this height adjustment
	var continueLanding = aircraft.Info.LandWhenIdle && self.CurrentActivity.IsCanceling && self.CurrentActivity.NextActivity == null;

	if (!continueLanding)
	{
		var dat = self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition);
		if (dat > aircraft.LandAltitude && dat < aircraft.Info.CruiseAltitude)
		{
			QueueChild(self, new TakeOff(self), true);
			return this;
		}

		aircraft.RemoveInfluence();
		return NextActivity;
	}
}
```

We'll also want something similar (but to land) in `TakeOff`, but can be left to a followup PR.

File: OpenRA.Mods.Common/Activities/Air/Fly.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Fly.cs]: We also need to cover `dat > aircraft.Info.CruiseAltitude` for aircraft that have just flown off the edge of a cliff!

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/Fly.cs]: We can fix this and IMO make the logic a bit clearer with:
```csharp
var dat = self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition);
if (IsCanceling)
{
	// We must return the actor to a sensible height before continuing
	// If the aircraft is on the ground we queue TakeOff to manage the influence reservation and takeoff sounds etc
	// TODO: It would be better to not take off at all, but we lack the plumbing to detect current airborne/landed state
	// If the aircraft lands when idle and is idle, we let the default idle handler manage this
	// TODO: Allow specific NextActivities to opt out of this height adjustment
	var skipHeightAdjustment = aircraft.Info.LandWhenIdle && self.CurrentActivity.IsCanceling && self.CurrentActivity.NextActivity == null;
	if (!skipHeightAdjustment && dat != aircraft.Info.CruiseAltitude)
	{
		if (dat <= aircraft.LandAltitude)
			QueueChild(self, new TakeOff(self, target), true);
		else
			VerticalTakeOffOrLandTick(self, aircraft, aircraft.Facing, aircraft.Info.CruiseAltitude);

		return this;
	}

	return NextActivity;
}
```

Note that i'm using `self.CurrentActivity.NextActivity` here and in `Land` rather than `NextActivity` because `NextActivity` will give the wrong result (usually `null`) if `Land` is queued as a child.

File: OpenRA.Mods.Common/Activities/Air/Land.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: Generally, they don't. They would if we queue `TakeOff` in `Fly.OnFirstRun` whenever the altitude is lower than `Aircraft.CruiseAltitude`, as I already suggested.

The only aircraft related activity that takes care of this itself, as far as I can see, is `HeliAttack` and it really shouldn't. All these activities should just queue `Fly` instead of calling `Fly.FlyTick` or any other internal `Fly` method directly. `Fly` would then queue `TakeOff` as necessary.

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: Makes sense. Replacing `FlyTick` with `Fly` children is going to be important for the acceleration and cliff height prediction logic too.

The activities that queue `Fly` directly or use `move.Move*` will then be able to tag themselves with `IActivityAircraftAltitudeAware` (or whatever we want to call it) to opt out of the internal `Land` handling.

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: In hindsight, my request to allow facing changes has caused more trouble than it's worth. In practice it also looks a bit odd to have units turn before landing on a helipad but then turn while taking off when reloading is complete.

How about we abandon this idea and keep it simple?

Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: I think units should be able to turn while landing as well.

 Since we will probably want to replace `FlyTick` with `Fly` children anyway and `Fly` is already queueing `TakeOff` (but in the wrong place), this _is_ keeping it simple.

Also, since any activity that cares about positioning at all will be queueing either `Fly` directly or using the `IMove` interface, we will be better off making the the "`TakeOff` on cancelling `Land`" thing an opt-in instead of an opt-out.

 (In fact, I've got the sneaking suspicion we will be left with exactly zero activities that would actually benefit from this, so we can just leave it for the idling case, making it even more simple.)

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: > I think units should be able to turn while landing as well.

Turning when landing is more complicated because we need to make sure that the facing is correct before it touches down, which may mean turning part way before starting to descend.

> Also, since any activity that cares about positioning at all will be queueing either Fly directly or using the IMove interface, we will be better off making the "TakeOff on cancelling Land" thing an opt-in instead of an opt-out.

Pragmatically, I expect you're correct. I'm just a bit annoyed that we still have to hardcode exceptions for aircraft in places like `Transforms`.

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: My main motivation for suggesting we drop the facing changes during takeoff is that the fly code is too broken to tackle in one PR, and each time through review we seem to find another direction or set of issues to scope creep it on. This is an area where we can at least keep the previous behaviour without introducing anything new to complain about, then the behaviour on both takeoff and land can be addressed together in a later PR.

**Comment by reaperrr** [comment]
Updated with the two suggested code parts by @pchote.

Not sure what else I should change right now.

File: OpenRA.Mods.Common/Activities/Air/Land.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: I agree that endlessly debating this isn't getting us anywhere. We should probably leave this as is and get this and #16509 merged quickly, so we can tackle these loose ends in one PR. Throwing away the facing changes during take-off would be a mistake though. Better to keep what we have here, so we have something to work from and polish later.

**Comment by pchote** [comment]
> Not sure what else I should change right now.

This is going to depend on the end goal that we want to reach. Based on the discussion with @tovl above my votes are for:
* Defering the facing adjustments during takeoff/landing to a future PR (possibly in Next + 1)
* Abandoning my suggestion for an `IActivityAircraftAltitudeAware` interface, and instead require activities to deal with arbitrary starting aircraft altitudes.
* Using `Aircraft.OnBecomingIdle` to do the final altitude fix for idle units

The first point is easy enough to address in this PR. The second and third ones rely on changes well out of scope for this PR, so changing "// TODO: Allow specific NextActivities to opt out of this height adjustment." comment to "// TODO: Remove this fixing all activities to work properly with arbitrary starting altitudes"

File: OpenRA.Mods.Common/Activities/Air/Land.cs
Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: I've been assuming that turning off the facing adjustment can be done by removing the `var desiredFacing = delta.HorizontalLengthSquared != 0 ? delta.Yaw.Facing : aircraft.Facing;` line and calling `VerticalTakeOffOrLandTick` with `aircraft.Facing` instead. This can be trivially restored in the followup.

Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: You're right. If it is just the one line, then that isn't a big deal.

**Comment by reaperrr** [comment]
> The first point is easy enough to address in this PR. The second and third ones rely on changes well out of scope for this PR, so changing "// TODO: Allow specific NextActivities to opt out of this height adjustment." comment to "// TODO: Remove this after fixing all activities to work properly with arbitrary starting altitudes" is hopefully sufficient.

Done.

File: OpenRA.Mods.Common/Activities/Air/Fly.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Fly.cs]: Add `aircraft.Info.CanHover &&` here to fix Banshees/Orca Bombers doing a VTOL step when stopped/issued a new order just after a cliff height change.

**Review by pchote — DISMISSED** [review]
I think we've covered all the major issues here now. Just one last simple fixup, then I think any other issues need to be deferred to followup PRs.

**Comment by reaperrr** [comment]
Fixed.

**Review by ghost — CHANGES_REQUESTED** [review]
When you build a carryall from a helipad in ts, other helicopters can no longer land on this helipad. This is not the case on bleed.

The same applies to the transport helicopters in RA and TD.

**Comment by reaperrr** [comment]
Updated, dropped the speculative fix for #16579 as that's what caused the issue @matjaeck reported above. Let's leave that for a dedicated follow-up as well.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: `during horizontal during`?

**Comment by reaperrr** [comment]
Fixed.

**Review by pchote — DISMISSED** [review]

File: OpenRA.Mods.Common/Activities/Air/TakeOff.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Air/TakeOff.cs]: Is this save to do? We had issues before with not doing such in `OnFirstRun`.

**Review by abcdefg30 — COMMENTED** [review]
There is a regression from this PR: If a hind runs out of ammo while (force) attacking, it returns to resupply. On bleed it proceeds to attack after rearming, with this PR it first moves to the rally point and then moves to attack again.

I'm not sure how critical this regression is. It could be better to open an issue and ensure it is fixed after all the aircraft and activity rewrites are done rather than holding this PR up.

File: OpenRA.Mods.Common/Activities/Air/TakeOff.cs
Comment (abcdefg30) [reply OpenRA.Mods.Common/Activities/Air/TakeOff.cs]: For the record:
```
02:04:08 (abcdefg30) so you agree that we can merge the PR once the OnFirstRun issue is solved?
02:04:35 (abcdefg30) I am kind of sure you can provoke breakage with that stuff being calculated in the ctor, but didn't construct a case (yet)
02:05:47 (pchote) yeah
02:06:00 (pchote) as much as i want to see the end of that PR, doing that in the ctor is bad
02:06:17 (pchote) but then lets merge ASAP once its fixed
```

Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Air/TakeOff.cs]: I just tried moving it to OnFirstRun locally, now the activity is broken in such a ridiculous way I don't even know how to debug it.
Pushed the changes so someone can take a look, but if nobody manages to debug that quickly, please let's settle for leaving it in the ctor and add a TODO for now.

**Comment by reaperrr** [comment]
> It could be better to open an issue and ensure it is fixed after all the aircraft and activity rewrites are done rather than holding this PR up.

Please do. This PR is bottlenecking my entire progress. All the fixes and follow-ups can mostly be filed and reviewed pretty much in parallel, but this PR is holding up all of that.

**Comment by pchote** [comment]
Fixed.

File: OpenRA.Mods.Common/Activities/Air/TakeOff.cs
Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/TakeOff.cs]: > I just tried moving it to OnFirstRun locally, now the activity is broken in such a ridiculous way I don't even know how to debug it.

I debugged this by `Console.WriteLine`ing the `rallyPointDestination` set in `OnFirstRun`.

The first run sets the correct value, but it failed and set `self.Location` the second time.
Wait, what? `OnFirstRun` should not run twice... Further investigation showed that `aircraft.UnReserve` was unconditionally queuing a fresh `TakeOff` that then broke things.

Adding a flag to `aircraft.UnReserve` to disable this fixed the problem.

**Review by pchote — DISMISSED** [review]

**Review by abcdefg30 — DISMISSED** [review]
For some reason Apaches now stack onto each other.

**Comment by reaperrr** [comment]
> For some reason Apaches now stack onto each other.

from IRC: > [22:17] <+reaperrr> the problem is that https://github.com/OpenRA/OpenRA/pull/16584/commits/2dddc03fd2e7343083fede2a59a90cc337a0512e#diff-558171f0691cfeefaa697d6b9dd0734bR174 doesn't make sure that delta contains no non-zero Z component

Fixed.

**Review by abcdefg30 — COMMENTED** [review]
I think this regresses airfield reservations. Testcase: Start the first Soviet mission and keep the planes in the air. Spend a bit of ammo on two of them and send the first to rearm. After that plane is rearmed, let it take off and send the second plane to the same airfield. It can't land.

**Comment by pchote** [comment]
I'm not able to repro the issue. Soviet01 disables the airfield rally point, so the plane will not take off automatically if ordered to return. It takes off automatically if you queue attack orders or force-attack the ground and wait for it to run out of ammo, but correctly releases the reservation in this case.

**Comment by pchote** [comment]
Still can't reproduce, after clarifying that "let it take off" actually means "order it to take off".

Instead, I found a different issue: If you let the yak completely run out of ammo and then try to order it to move before it has fully loaded it will freeze on the airfield and other planes can land on top of it.

**Comment by pchote** [comment]
```
21:25 <+pchote> now i can't get the planes to land at all because they are being repulsed out of the map
21:25 <+abcdefg30> are you sure
21:25 <+abcdefg30> that sounds what I was experiencing
21:25 <+abcdefg30> move them back inside
21:26 <+abcdefg30> and try to let them land
21:26 <+abcdefg30> (again from inside)
21:26 <+abcdefg30> because the airfields are at the border, the planes idle outside of the map
21:26 <+pchote> oh
21:26 <+pchote> more english fail
21:26 <+pchote> i thought you were saying that the reservation was blocked
21:26 <+pchote> so you don't get the land cursor
21:26 <+abcdefg30> oh fail
21:26 <+abcdefg30> sorry
21:27 <+abcdefg30> yes you get the land cursor but they don't land
21:27 <+abcdefg30> unless you let the original yak land on another airfield iirc
21:27 <+abcdefg30> (which led me to believe reservations are borked)
```

**Comment by matjaeck** [comment]
> I think this regresses airfield reservations. Testcase: Start the first Soviet mission and keep the planes in the air. Spend a bit of ammo on two of them and send the first to rearm. After that plane is rearmed, let it take off and send the second plane to the same airfield. It can't land.

See reproduction below.

<details>

takeoffpr

I clicked on the airfield with the "blocked" cursor before the plane landed.

**Comment by pchote** [comment]
The second yak holds the `Reservable` reservation on the airfield, but is unable to land. I'm pretty sure this means that the first yak did not remove its cell reservation in the `ActorMap`.

**Comment by pchote** [comment]
Ok, this is an interesting one. Something cancels the `TakeOff` activity before it can first tick. I haven't debugged deep enough to find out exactly what, yet.

Canceling the activity changes its `.State` from `Queued` to `Canceling`, which causes `Activity.TickOuter` to completely skip `OnFirstRun`! This is a fairly serious activity plumbing bug that we'll need to fix - maybe in #16481, or maybe in its own PR.

`TakeOff.Tick` doesn't check for `IsCanceling`, so one option could be to set `IsInterruptable = false`. The problem here is that this would also make the attack-move-to-rallypoint uninterruptable, which may be just as bad?

**Comment by matjaeck** [comment]
> The problem here is that this would also make the attack-move-to-rallypoint uninterruptable, which may be just as bad?

Yes, this won't work from a gameplay perspective. Could we move the attack-move-to-rallypoint part out of the takeoff activity?

**Comment by pchote** [comment]
This is another double-takeoff issue, the same as the rally point one earlier.

Adding `skipTakeOff: true` to the `Unreserve` calls in `Aircraft`'s Move/Enter/Repair order handling fixes the issue.

We'll need to check all other `Unreserve` calls and make sure that any other cases which immediately queue their own fly/takeoff/etc also set the flag.

**Comment by reaperrr** [comment]
> Canceling the activity changes its .State from Queued to Canceling, which causes Activity.TickOuter to completely skip OnFirstRun! This is a fairly serious activity plumbing bug that we'll need to fix - maybe in #16481, or maybe in its own PR.

FWIW, that was a known issue to me and mentioned by @obrakmann before as well, because that very bug was the reason we moved the re-queueing of unfinished `Resupply` from `ResolveOrder` to `OnBecomingIdle`. I thought we had a ticket for this, but I guess with all the changes around the activity plumbing I lost track of what bugs were known, what bugs were already fixed etc.

**Comment by pchote** [comment]
Interesting. I either had no idea or completely forgot about it.

With the activity rework pushing more important things into `OnFirstRun` this has become a critical issue to fix before the playtest, but certainly not here.

**Comment by pchote** [comment]
The fix (5b15387d24e2e488716b0444fefaf9746994a1b5) is simple, so I can file a PR with this and commits to undo the `skipTakeOff: true` changes here and maybe the `Resupply` change after this one is merged.

**Comment by reaperrr** [comment]
> Adding skipTakeOff: true to the Unreserve calls in Aircraft's Move/Enter/Repair/ReturnToBase order handling and Aircraft.Nudge should fix this and the other issues that we didn't discover during ingame testing.

Updated, except I approached it the other way round: I turned the take-off into an opt-in.
There are only 2 situations where we actually want that automatic take-off:
- if `MayYieldReservation` is `true` and some other aircraft requests reservation
- if the reserved actor becomes invalid (destroyed, sold etc.)
So making it opt-in considerably reduced the diff, as there are just 2 places in `Reservable` that need it set to `true` now.

**Review by ghost — APPROVED** [review]
LGTM.

**Review by pchote — APPROVED** [review]
This has been blocking progress for too long, so lets push any further regression fixes to followup PRs.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/609c5a8738b1bc24e6f78b30118273dd234d29d6

**Pull request merged**
